### PR TITLE
Event streaming

### PR DESCRIPTION
**Pull request by blinky3713**
The point of this pr is to enable more flexible event streaming. currently it is the case that filters are created, but the application developer does not have access to them directly. As a result, one can only poll for events from `Latest` to `Latest`. Here we take some ideas developed and tested in purescript-web3, using kmett's [machines](https://hackage.haskell.org/package/machines) library to set up the streams. In short here are the changes:

1. change the `Filter` object to require the `filterToBlock` and `filterFromBlock`, rather than a `Maybe`. The fact that you can get away with using the defaults if the filter fields are not specified is not a good reason to be lazy, and it made dealing with events in the rest of these changes way harder and more ad-hoc.

2. The goal is to allow for seamlessly streaming events from an arbitrary `DefaultBlock` to an arbitrary 'DefaultBlock'. Take the following examples. Say we would like to stream 'Transfer' events from an ERC20 contract, and we are currently at block 15.

Stream events from block 10 to block 20. The event streamer should use `getLogs` to fetch the first blocks 10 ... 15, then create a new filter and use `getFilterChanges` for blocks 15 ... 20 and perform the appropriate clean up.

We would like to stream events from block 0 to block 10. We should use `getLogs` for everything.

We would only like to stream Latest to Latest, the streamer should make a new filter, then call `getFilterChanges` for everything.

We would like to stream 18 to Latest. It should make a new filter and call eth_getFilterChanges for everything.

Furthermore, we would like the option of specifying a range of blocks for anything which uses `getLogs` rather than going block by block. You should be able to derive the normal case by specifying the range to be 1. 

Let me know what you think. Please see the tests for specific edge cases

### CODE REVIEW

**Comment by akru** [comment]
@blinky3713 Thank you, it's very interesting improvement.
Can you explain it in code examples `with` and `without`?

For example, simple event subscription is:
```haskell
event tokenAddress $ \Transfer f t v -> do
  liftIO (print f)
```

**Comment by blinky3713** [comment]
i'm not sure i understand what you mean by `with` and `without`, maybe an example of how things look now would help though.

```haskell
  let baseFilter = eventFilter (Proxy :: Proxy Transfer) erc20Address) 
  allTimeFilter = baseFilter { filterFromBlock = Earliest }
   forkWeb3 $ event allTimeFilter $ \(Transfer{...} ->
      liftIO . print $ transferTo 
```

two main differences.
1. you construct the filter object yourself rather than relying on the `event` function to figure it out for you.
2. `event` does not automatically fork a thread anymore, you need to do that by hand. And actually, `forkWeb3` no longer uses a simple `forkIO` but `async` from `Control.Concurrent.Async`. This is strictly more general than `forkIO` and allows for much finer control -- for rethrowing exceptions in the main thread, for waiting on a thread to finish before proceeding, etc.

you can see the tests for more examples

**Comment by blinky3713** [comment]
one interesting improvement could be that `eventFilter` returns a tagged filter, ie.

```haskell
class Event e where
  eventFilter ::   eventFilter :: Proxy e -> Address -> Tagged e Filter

event :: forall p i ni e .
         ( Provider p
         , DecodeEvent i ni e
         , Event e
         )
      => Tagged e Filter
      -> (e -> ReaderT Change (Web3 p) EventAction)
      -> Web3 p ()
```

so that you cannot accidentally mix filters. However, this would probably require making lenses of the form

```haskell
_toBlock :: Lens' (Tagged e Filter) DefaultBlock
...
```

in order to facilitate customizing the filters.

**Comment by akru** [comment]
Tagged filters looks good, cause you shouldn't double set event type: in filter proxy decl and in event handler. So, may be filter with phantom type like `data Filter e = Filter...` is more simple way.

I have two proposals:

1. Tagged/phantom event filter datatype;
2. Threaded `event` function by default (in my opinion, async start of event is more frequently case), for blocking call can be declarated special function like `event'`, and async call can be declared as `event a b = forkWeb3 $ event' a b`.

**Comment by blinky3713** [comment]
ok made the changes,

- put in a phantom type for event filters
- made event async by default

File: src/Network/Ethereum/Web3/Contract.hs [hunk 1]
```diff
@@ -78,44 +93,191 @@ data EventAction = ContinueEvent
 -- | Contract event listener
 class Event e where
     -- | Event filter structure used by low-level subscription methods
-    eventFilter :: Proxy e -> Address -> Filter
+    eventFilter :: Proxy e -> Address -> Filter e
```
Comment (akru) [inline src/Network/Ethereum/Web3/Contract.hs]: Is `Proxy e` really needed? Because `e` typevar already set at `Filter e`.

**Comment by akru** [comment]
Also, it will be great to create broadcast event filters to receive events from all smart contracts.

My proposal is create multi-parameter typeclass:

```haskell
class Event a e where
  eventFilter :: a -> Filter e
```

and instance for common event cases like

```haskell
instance Event () e where ...
-- ^ Broadcast event listener (listen event from all contracts)
instance Event Address e where ...
-- ^ Unicast event listener (listen event from address)
instance Event [Address] e where ...
-- ^ Multiple address event listener
```

and also will be interested event conversion instance:

```haskell
instance Event a b => Event (Filter b) e where ...
```

**Comment by blinky3713** [comment]
i removed the proxy arg.

As for the other, i think it belongs in a separate github issue. All I'm really hoping to do here is open up the use of playing events from arbitrary points in time and it seems like this feature is complete.

**Comment by akru** [comment]
Ok, I’m agree
